core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs

```diff
@@ -242,22 +243,34 @@ impl<G: L1GasPriceProvider> EthNamespace<G> {
         const METHOD_NAME: &str = "get_filter_logs";
 
         let method_latency = API_METRICS.start_call(METHOD_NAME);
-        // Note: We have to keep this as a separate variable, since otherwise the lock guard would exist
-        // for duration of the whole `match` block, and this guard is not `Send`. This would make the whole future
-        // not `Send`, since `match` has an `await` point.
-        let maybe_filter = self.state.installed_filters.read().await.get(idx).cloned();
-        let filter = match maybe_filter {
-            Some(TypedFilter::Events(filter, _)) => {
-                let from_block = self
-                    .state
-                    .resolve_filter_block_number(filter.from_block)
-                    .await?;
-                TypedFilter::Events(filter, from_block)
-            }
-            _ => return Err(Web3Error::FilterNotFound),
+        // We clone the filter to not hold the filter lock for an extended period of time.
+        let maybe_filter = self
+            .state
+            .installed_filters
+            .lock()
+            .await
+            .get_and_update_stats(idx)
+            .cloned();
+
+        let Some(TypedFilter::Events(filter, _)) = maybe_filter else {
+            return Err(Web3Error::FilterNotFound);
         };
 
-        let logs = self.filter_changes(filter).await?.0;
+        let from_block = self
+            .state
+            .resolve_filter_block_number(filter.from_block)
+            .await?;
+
+        let mut typed_filter = TypedFilter::Events(filter, from_block);
+
+        let logs = self.filter_changes(&mut typed_filter).await?;
+
+        self.state
+            .installed_filters
+            .lock()
+            .await
+            .update(idx, typed_filter);
```

Comment:
Eh, isn't the previous behavior (the filter is **not** updated by the call) the correct one? Maybe, I wasn't clear in the previous comment here – I wanted you to add a comment why the filter shouldn't be updated, not to add logic for updating the filter.

Reply:
Uh, yes, I guess did not understand you correctly